### PR TITLE
speed up sync fetch by contract ID/key with postgres and faster in-memory filtering

### LINKED ISSUE #7711: Use the DB backend for all synchronous query endpoints in the JSON API
I don’t really see a reason for why we shouldn’t use the postgres backend for `fetch` and `fetchByKey` so let’s change that to be more consistent and improve performance.

### PR DESCRIPTION
**Pull request by S11001001**
```rst
CHANGELOG_BEGIN
- [JSON API] ``/v1/fetch`` now uses the Postgres database, if configured, to
  look up contracts by ID or key, except when querying a contract by ID without
  its corresponding template ID.  The fallback in-memory version of
  ``/v1/fetch`` is also now significantly more efficient for large datasets,
  though still linear.

  You may optionally re-create JSON API's database to take full advantage.
  See `issue #7993 <https://github.com/digital-asset/daml/pull/7993>`__.
CHANGELOG_END
```

Additionally, we

* move towards a separated DB dao and in-memory dao with the `Search` type
* shift the filtering for in-memory lookups to the other side of the natural transformation, so the time and memory consumption comparison is much fairer

Fixes #7711.

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger-service/db-backend/src/main/scala/com/digitalasset/http/dbbackend/Queries.scala [hunk 1]
```diff
@@ -244,6 +244,23 @@ object Queries {
...
 
+  private[http] def fetchById(
+      parties: OneAnd[Set, String],
+      tpid: SurrogateTpId,
+      contractId: String)(
+      implicit log: LogHandler,
+      gvs: Get[Vector[String]],
+      pvs: Put[Vector[String]])
+    : ConnectionIO[Option[DBContract[Unit, JsValue, JsValue, Vector[String]]]] =
+    selectContracts(parties, tpid, sql"contract_id = $contractId").option
```
Comment (cocreature) [inline ledger-service/db-backend/src/main/scala/com/digitalasset/http/dbbackend/Queries.scala:255]: Should we create indices on `contract_id` and maybe `key` by default? Seems common enough that we might not want to leave this to users.

File: ledger-service/http-json/src/main/scala/com/digitalasset/http/ContractsService.scala [hunk 2]
```diff
@@ -79,60 +84,85 @@ class ContractsService(
...
+          .runWith(Sink.headOption): Future[Option[Error \/ domain.ActiveContract[LfValue]]]
+
+        result <- lookupResult(errorOrAc)
+
+      } yield result
+    }
+
+    override def search(ctx: SearchContext[Set, Id], queryParams: Map[String, JsValue]) = {
+      import ctx.{jwt, parties, templateIds}
+      searchInMemory(jwt, parties, templateIds, InMemoryQuery.Params(queryParams))
```
Comment (cocreature) [inline ledger-service/http-json/src/main/scala/com/digitalasset/http/ContractsService.scala:163]: This didn’t change in this PR but why are we representing queries as `Map[String, JsValue]`. This seems like we are encoding the fact that the first layer is a record? Is that useful given that we lose all structure on the levels below and therefore have to deal with incorrectly typed queries anyway?

File: ledger-service/http-json/src/main/scala/com/digitalasset/http/ContractsService.scala [hunk 3]
```diff
@@ -184,69 +214,115 @@ class ContractsService(
         status = StatusCodes.BadRequest
       )
     } else {
-      val source = daoAndFetch.cata(
-        x => searchDb(x._1, x._2)(jwt, parties, resolvedTemplateIds, queryParams),
-        searchInMemory(jwt, parties, resolvedTemplateIds, queryParams)
+      val searchCtx = SearchContext[Set, Id](jwt, parties, resolvedTemplateIds)
+      val source = SearchDb.cata(
```
Comment (cocreature) [inline ledger-service/http-json/src/main/scala/com/digitalasset/http/ContractsService.scala:218]: Can we use `search.search` instead? If we have an abstraction for the backend we should also use it and not pattern match on the backend here again.

**Review by cocreature — APPROVED** [review]
Nice work, thanks a lot!

File: ledger-service/http-json/src/main/scala/com/digitalasset/http/ContractsService.scala (hunk 2 above)
Comment (S11001001) [reply ledger-service/http-json/src/main/scala/com/digitalasset/http/ContractsService.scala:163]: It's been a while, but as I vaguely recall, certain invariants are easier to enforce if you type this fact. And all else being equal I'm loath to give up a carefully typed premise.

File: ledger-service/db-backend/src/main/scala/com/digitalasset/http/dbbackend/Queries.scala (hunk 1 above)
Comment (S11001001) [reply ledger-service/db-backend/src/main/scala/com/digitalasset/http/dbbackend/Queries.scala:260]: `contract_id` is already a primary key. For contract keys, 18e80597c7a74d0a311b621ae3d63fbe6684b2a7

File: ledger-service/http-json/src/main/scala/com/digitalasset/http/ContractsService.scala (hunk 3 above)
Comment (S11001001) [reply ledger-service/http-json/src/main/scala/com/digitalasset/http/ContractsService.scala]: a689cf6552e

**Pull request merged**